scipy/optimize/cython_optimize/zeros.pyx

```diff
@@ -0,0 +1,70 @@
+from . cimport c_zeros
+
+
+# callback function wrapper that extracts function, args from params struct
+cdef double scipy_zeros_functions_func(double x, void *params):
+    cdef scipy_zeros_parameters *myparams = <scipy_zeros_parameters *> params
+    cdef void* args = myparams.args
+    cdef callback_type f = myparams.function
+
+    return f(x, args)  # callback_type takes a double and a struct
+
+
+# cythonized way to call scalar bisect
+cdef double bisect(callback_type f, double xa, double xb, void *args, double xtol, double rtol, int iter, scipy_zeros_parameters *full_output):
+    cdef double root
+    cdef scipy_zeros_parameters myparams
+    # create params struct
+    myparams.args = args
+    myparams.function = f
+    root = c_zeros.bisect(scipy_zeros_functions_func, xa, xb, xtol, rtol, iter, <c_zeros.default_parameters *> &myparams)
+    if full_output is not NULL:
```

Comment:
seems unfortunate that this code block has to be duplicated in each of the functions

Reply:
>seems unfortunate that this code block has to be duplicated in each of the functions

I agree, @tylerjereddy , and I started [another branch to use templates to generate the code automatically](../../../mikofski/scipy/tree/cython_optimize_generate_code/scipy/optimize/cython_optimize/templates), but I thought it would be easier to review if the templates were implemented in a separate PR. Note, when I started the `cython_optimize_generate_code` branch I used Jinja2 templates, but then based on [this thread in SciPy-dev mailing list](https://mail.python.org/pipermail/scipy-dev/2018-August/023012.html), I learned about Tempita, although I haven't implemented the change yet.

Reply:
✔️ done